### PR TITLE
[SPARK-25499][TEST]Refactor BenchmarkBase and Benchmark

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

Currently there are two classes with the same naming BenchmarkBase:
1. `org.apache.spark.util.BenchmarkBase`
2. `org.apache.spark.sql.execution.benchmark.BenchmarkBase`

This is very confusing. And the benchmark object `org.apache.spark.sql.execution.benchmark.FilterPushdownBenchmark` is using the one in `org.apache.spark.util.BenchmarkBase`, while there is another class `BenchmarkBase` in the same package of it...

Here I propose:
1. the package `org.apache.spark.util.BenchmarkBase` should be in test package of core module. Move it to package `org.apache.spark.benchmark` .
2. Move `org.apache.spark.util.Benchmark` to test package of core module. Move it to package `org.apache.spark.benchmark` .
3. Rename the class `org.apache.spark.sql.execution.benchmark.BenchmarkBase` as `BenchmarkWithCodegen`

## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
@wangyum @yucai @dongjoon-hyun @cloud-fan @gatorsmile 
Let's focus on this before we merge other benchmark PRs.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala [hunk 1]
```diff
@@ -27,7 +27,7 @@ import org.apache.spark.sql.functions.monotonically_increasing_id
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.internal.SQLConf.ParquetOutputTimestampType
 import org.apache.spark.sql.types.{ByteType, Decimal, DecimalType, TimestampType}
-import org.apache.spark.util.{Benchmark, BenchmarkBase => FileBenchmarkBase, Utils}
+import org.apache.spark.util.Utils
 
 /**
  * Benchmark to measure read performance with Filter pushdown.
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala]: How about change scala doc to below to fix **fails to generate documentation**?
```scala
 * To run this benchmark:
 * {{{
 *   1. without sbt: bin/spark-submit --class <this class> <spark sql test jar>
 *   2. build/sbt "sql/test:runMain <this class>"
 *   3. generate result: SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain <this class>"
 *      Results will be written to "benchmarks/FilterPushdownBenchmark-results.txt".
 * }}}
```
fails to generate documentation error message:
```java
/home/jenkins/workspace/SparkPullRequestBuilder@2/target/javaunidoc/org/apache/spark/mllib/linalg/UDTSerializationBenchmark.html...
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/mllib/target/java/org/apache/spark/mllib/linalg/UDTSerializationBenchmark.java:5: error: unknown tag: this
[error]  * 1. without sbt: bin/spark-submit --class <this class> <spark mllib test jar>
[error]                                             ^
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/mllib/target/java/org/apache/spark/mllib/linalg/UDTSerializationBenchmark.java:5: error: unknown tag: spark
[error]  * 1. without sbt: bin/spark-submit --class <this class> <spark mllib test jar>
[error]                                                          ^
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/mllib/target/java/org/apache/spark/mllib/linalg/UDTSerializationBenchmark.java:6: error: unknown tag: this
[error]  * 2. build/sbt "mllib/test:runMain <this class>"
[error]                                     ^
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/mllib/target/java/org/apache/spark/mllib/linalg/UDTSerializationBenchmark.java:7: error: unknown tag: this
[error]  * 3. generate result: SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "mllib/test:runMain <this class>"
[error] 
```

**Comment by wangyum** [comment]
`KryoBenchmark` is in core, and `UnsafeProjectionBenchmark`, `HashByteArrayBenchmark`  and `HashBenchmark` are in `catalyst`. If we move the benchmark base class to sql, benchmarks mentioned above would not be able to inherit from the benchmark base class. What do you think?

**Comment by gengliangwang** [comment]
> KryoBenchmark is in core, and UnsafeProjectionBenchmark, HashByteArrayBenchmark and HashBenchmark are in catalyst. If we move the benchmark base class to sql, benchmarks mentioned above would not be able to inherit from the benchmark base class. What do you think?

@wangyum The cases you mentioned are currently using the `org.apache.spark.sql.execution.benchmark.BenchmarkBase` (the old one), so it seems fine.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/FilterPushdownBenchmark.scala]: Thanks, I have updated the doc 👍

**Comment by gengliangwang** [comment]
retest this please.

File: core/src/main/scala/org/apache/spark/sql/execution/benchmark/Benchmark.scala [hunk 2]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.apache.spark.util
+package org.apache.spark.sql.execution.benchmark
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/sql/execution/benchmark/Benchmark.scala]: this class is in core and we should not have `sql` in the package name

File: core/src/main/scala/org/apache/spark/sql/execution/benchmark/BenchmarkBase.scala [hunk 3]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.apache.spark.util
+package org.apache.spark.sql.execution.benchmark
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/sql/execution/benchmark/BenchmarkBase.scala]: ditto

**Comment by cloud-fan** [comment]
Please also explain which module(core or sql?) these benchmark classes should be, in the PR description.

**Comment by gengliangwang** [comment]
@cloud-fan @wangyum  Thanks for the suggestion. I have updated the target package and related PR description.

**Comment by cloud-fan** [comment]
LGTM

**Comment by dongjoon-hyun** [comment]
+1, late LGTM.

**Pull request closed without merging**